test/common/grpc/codec_test.cc

```diff
@@ -106,6 +107,41 @@ TEST(GrpcCodecTest, decodeInvalidFrame) {
   EXPECT_EQ(size, buffer.length());
 }
 
+// A frame of nothing followed by a frame of 0's followed then invalid
+TEST(GrpcCodecTest, decodeNullByteFrameWithInvalidFrameAfterward) {
+  const std::string data("\000\000\000\000\0000000", 9);
+  Buffer::OwnedImpl buffer(data.data(), data.size());
+  std::vector<Frame> frames;
+  Decoder decoder;
+  ENVOY_LOG_MISC(trace, "Buffer: {}", buffer.toString());
+  EXPECT_FALSE(decoder.decode(buffer, frames));
+  // When the decoder doesn't successfully decode, it puts valid frames up until
+  // an invalid frame into output frame vector.
+  EXPECT_NE(0, frames.size());
+}
+
+// If there is valid frames followed by an invalid frame, the decoder will succesfully put the valid
+// frame in the output and return false due to the invalid frame
+TEST(GrpcCodecTest, decodeValidFramesWithInvalidFrameAfterward) {
+  // Decode a valid encoded structured response plus invalid data afterward
+  grpc::health::v1::HealthCheckResponse response;
+  response.set_status(grpc::health::v1::HealthCheckResponse::SERVING);
+  const auto data = Grpc::Common::serializeToGrpcFrame(response);
+  std::vector<uint8_t> buffer_vector = std::vector<uint8_t>(data->length(), 0);
+  data->copyOut(0, data->length(), &buffer_vector[0]);
+  for (size_t i = 0; i < 6; i++) {
+    buffer_vector.push_back(48); // Represents ASCII Character of 0
+  }
+  Buffer::OwnedImpl buffer(buffer_vector.data(), buffer_vector.size());
+  std::vector<Frame> frames;
+  Decoder decoder;
+  ENVOY_LOG_MISC(trace, "Buffer: {}", buffer.toString());
+  EXPECT_FALSE(decoder.decode(buffer, frames));
+  // When the decoder doesn't successfully decode, it puts valid frames up until
```

Comment:
is the output buffer only the first frame's size? would also be good to check.

Reply:
Sorry, I'm not really sure what you mean by this. https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/codec.cc#L41 drains the buffer only on a succesful decode. Thus, since there's an invalid frame at the end of buffer, buffer length != first frame length right? I added a NE check for NE

Reply:
// Buffer does not get drained due to it returning false.
  EXPECT_EQ(size, buffer.length());
  // Only part of the buffer represented a valid frame. Thus, the frame length should not equal the buffer length.
  EXPECT_EQ(size, frames[0].length());

Reply:
That's what I think you mean?

Reply:
Awesome, added to both.

Reply:
Ahhh, yes, I forgot the input also gets drained! Thank you. buffer stays full, and at the same time frames is the size of the good frame. (should be smaller than data.size() though right? since the invalid part is ignored?)

Reply:
This is what it amounted to: // Buffer does not get drained due to it returning false.
  EXPECT_EQ(size, buffer.length());
  // Only part of the buffer represented a valid frame. Thus, the frame length should not equal the buffer length.
  EXPECT_NE(size, frames[0].length_); Had a few typos in the first one :P